### PR TITLE
feat: init cli

### PR DESCRIPTION
**Pull request by gao-sun**
## Summary

- the initial PR for package `cli`, including a TTY version installation CLI.
- inherits network proxy to download release from environment variables

### Up next
- replace `install.js` and test in Windows
- `npm create logto` or `npm logto init` compatibility
- `npm @logto/cli xxx` compatibility
- support more commands e.g. database alteration, move env to database

### Node 16

image

### Node 18

image

### No Postgres

image

### Path exists

image

### CODE REVIEW

File: packages/cli/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+import { downloadFile, log, safeExecSync } from './utilities';
+
+const pgRequired = new semver.SemVer('14.0.0');
+
+const validateNodeVersion = () => {
+  const required = new semver.SemVer('16.0.0');
+  const current = new semver.SemVer(execSync('node -v', { encoding: 'utf8', stdio: 'pipe' }));
+
+  if (required.compare(current) > 0) {
+    log.error(`Logto requires NodeJS >=${required.version}, but ${current.version} found.`);
```
Comment (simeng-li) [inline packages/cli/src/index.ts:22]: should we return here and end the cli process ?

**Review by simeng-li — APPROVED** [review]
overall LGTM

File: packages/cli/src/utilities.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  info: (...args) => {
+    console.log(chalk.blue('[info]'), ...args);
+  },
+  warn: (...args) => {
+    console.log(chalk.yellow('[warn]'), ...args);
+  },
+  error: (...args) => {
+    console.log(chalk.red('[error]'), ...args);
+    // eslint-disable-next-line unicorn/no-process-exit
+    process.exit(1);
```
Comment (simeng-li) [inline packages/cli/src/utilities.ts:31]: nit: prefer to keep it at the outer scope, so the logging behavior is clean.

**Review by charIeszhao — APPROVED** [review]
approved with comments

File: packages/cli/package.json [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  "name": "@logto/cli",
+  "version": "1.0.0-beta.10",
+  "description": "Logto CLI.",
+  "author": "Silverhand Inc. <contact@silverhand.io>",
+  "homepage": "https://github.com/logto-io/logto#readme",
+  "license": "MPL-2.0",
+  "main": "lib/index.js",
+  "bin": {
+    "logto": "bin/logto",
+    "lg": "bin/logto"
```
Comment (wangsijie) [inline packages/cli/package.json:11]: Is this an alias?

**Review by wangsijie — APPROVED** [review]

File: packages/cli/src/index.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (gao-sun) [reply packages/cli/src/index.ts:22]: the `log.error()` itself ends the process.

File: packages/cli/src/utilities.ts (hunk 2 above)
Comment (gao-sun) [reply packages/cli/src/utilities.ts:31]: in most cases it should end immediately. maybe we can provide another function like `log.message('error', ...args)` to achieve that?

File: packages/cli/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+import * as prompts from 'prompts';
+import * as semver from 'semver';
+import tar from 'tar';
+
+import { downloadFile, log, safeExecSync } from './utilities';
+
+const pgRequired = new semver.SemVer('14.0.0');
+
+const validateNodeVersion = () => {
+  const required = new semver.SemVer('16.0.0');
```
[the remark this answers is not in the document]
Comment (gao-sun) [reply packages/cli/src/index.ts:18]: it's already a constant? the `SemVer` class represents a version number.

File: packages/cli/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+const downloadRelease = async () => {
+  const tarFilePath = path.resolve(os.tmpdir(), './logto.tar.gz');
+
+  log.info(`Download Logto to ${tarFilePath}`);
+  await downloadFile(
+    'https://github.com/logto-io/logto/releases/latest/download/logto.tar.gz',
+    tarFilePath
+  );
+
+  return tarFilePath;
```
[the remark this answers is not in the document]
Comment (gao-sun) [reply packages/cli/src/index.ts:99]: if it fails the whole process fails. try catch is needed until we have retry strategy?

File: packages/cli/package.json (hunk 3 above)
Comment (gao-sun) [reply packages/cli/package.json:11]: yes

File: packages/cli/src/index.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (charIeszhao) [reply packages/cli/src/index.ts:99]: ok

**Comment by gao-sun** [comment]
i saw the comments are non-blocking, so i'm going to merge it first - we'll discuss offline and update in a separate PR if needed

File: packages/cli/package.json (hunk 3 above)
Comment (gao-sun) [reply packages/cli/package.json:11]: will remove

**Pull request merged**